---
title: "【Go】errors.Isを完全に理解した"
emoji: "🤯"
type: "tech" # tech: 技術記事 / idea: アイデア
topics: ["go"]
published: true
published_at: 2024-12-05 06:59
---

# はじめに

Goの`errors.Is`は、エラーの比較を行うための関数です。
でもこの関数がどういうケースで`true`を返すのか、いまいちよく分かってないという方もいらっしゃるのではないでしょうか？
~~私もその1人でした。~~
この記事では、問題とその解答を通して、`errors.Is`の挙動を完全に理解することを目指します。

# まずは問題

以下のQ1-Q7について、`errors.Is`関数が`true`を返すかどうかを考えてみましょう。

```go
package main

import (
	"errors"
	"fmt"
)

func main() {
	err1 := errors.New(ErrMsg)
	err2 := err1
	fmt.Printf("Q1: %v\n", errors.Is(err1, err2))

	err1 = errors.New(ErrMsg)
	err2 = errors.New(ErrMsg)
	fmt.Printf("Q2: %v\n", errors.Is(err1, err2))

	err1 = fmt.Errorf("wrapped: %w", ErrSentinel)
	err2 = ErrSentinel
	fmt.Printf("Q3: %v\n", errors.Is(err1, err2))

	err1 = fmt.Errorf("wrapped: %w", ErrSentinel)
	err2 = fmt.Errorf("wrapped: %w", ErrSentinel)
	fmt.Printf("Q4: %v\n", errors.Is(err1, err2))

	err1 = ErrSentinel
	err2 = fmt.Errorf("wrapped: %w", ErrSentinel)
	fmt.Printf("Q5: %v\n", errors.Is(err1, err2))

	err1 = &MyError{Msg: ErrMsg}
	err2 = errors.New(ErrMsg)
	fmt.Printf("Q6: %v\n", errors.Is(err1, err2))

	err1 = ErrSentinel
	err2 = &MyError{Msg: ErrMsg}
	fmt.Printf("Q7: %v\n", errors.Is(err1, err2))
}

const ErrMsg = "error message"

var ErrSentinel = errors.New(ErrMsg)

type MyError struct {
	Msg string
}

func (e *MyError) Error() string {
	return e.Msg
}

func (e *MyError) Is(target error) bool {
	return e.Msg == target.Error()
}
```

## version

```bash
$ go version
go version go1.23.3 darwin/arm64
```

# 解説と解答

先に解答を見たい方は[解答](#解答)へ。

## 先にまとめ

- `errors.Is`の挙動を理解したいなら、`errors`パッケージの`is`関数を見よ
- `is`関数では基本的にはポインタが一致してるかだけを見る
- 基本的にはエラーメッセージの中身を比較するような実装にはなっていない
- `errors.Is`の第一引数の`err`は`Unwrap`を繰り返され、第二引数とポインタが一致しているかが見られる
- `errors.Is`で`Unwrap`を繰り返されるのは第一引数の方であり、第二引数の`target`は`Unwrap`されない
- `errors.Is`で第一引数の`err`に`Is`メソッドがある場合は、その返り値で評価される
- `errors.Is`で`Is`メソッドが評価されるのは第一引数の方であり、第二引数の`target`の`Is`メソッドは評価されない

## Q1

```go
    err1 := errors.New(ErrMsg)
    err2 := err1
    fmt.Printf("Q1: %v\n", errors.Is(err1, err2))
```

`errors.Is`の挙動を理解するには、**`errors.Is`の実装を見るのが一番早いです。**

```go  
func Is(err, target error) bool {
	if target == nil {
		return err == target
	}

	isComparable := reflectlite.TypeOf(target).Comparable()
	return is(err, target, isComparable)
}
```

最初の3行は`target`が`nil`の場合を処理しているだけです。
`isComparable`は今回のケースでは全てComparable, すなわち`true`を返します。
ということで、**`is`関数が実質的な本体**です。

```go
func is(err, target error, targetComparable bool) bool {
	for {
		if targetComparable && err == target {
			return true
		}
		if x, ok := err.(interface{ Is(error) bool }); ok && x.Is(target) {
			return true
		}
		switch x := err.(type) {
		case interface{ Unwrap() error }:
			err = x.Unwrap()
			if err == nil {
				return false
			}
		case interface{ Unwrap() []error }:
			for _, err := range x.Unwrap() {
				if is(err, target, targetComparable) {
					return true
				}
			}
			return false
		default:
			return false
		}
	}
}
```

forについては後で説明するとして、最初の条件分岐を見ます。
`targetComparable`は`true`なので、基本的に`err == target`が評価されます。
これは**同じポインタであれば`true`を返す**ということです。

Q1では以下のように明らかに同じポインタを指しているので、`true`が返ります。

```go
    err1 := errors.New(ErrMsg)
    err2 := err1
    fmt.Printf("Q1: %v\n", errors.Is(err1, err2)) // true
```

## Q2

```go
    err1 = errors.New(ErrMsg)
    err2 = errors.New(ErrMsg)
    fmt.Printf("Q2: %v\n", errors.Is(err1, err2))
```

Q1から分かったように、`errors.Is`は基本的にはポインタの一致を見ます。
**エラーメッセージの中身を比較するような実装にはなっていません。**
`err1`と`err2`は同じエラーメッセージを持っているものの、それぞれ`errors.New`で生成されているので、異なるポインタを持っています。
よって、`false`が返ります。

```go
    err1 = errors.New(ErrMsg)
    err2 = errors.New(ErrMsg)
    fmt.Printf("Q2: %v\n", errors.Is(err1, err2)) // false
```

## Q3

```go
    err1 = fmt.Errorf("wrapped: %w", ErrSentinel)
    err2 = ErrSentinel
    fmt.Printf("Q3: %v\n", errors.Is(err1, err2))
```

もう一度`errors.is`の実装を覗いてみます。

```go
func is(err, target error, targetComparable bool) bool {
	for {
        // 中略
		switch x := err.(type) {
		case interface{ Unwrap() error }:
			err = x.Unwrap()
			if err == nil {
				return false
			}
        // 中略
		default:
			return false
		}
	}
}
```

以下の部分は要するに、「`err`が`error`を返す`Unwrap`メソッドを持っている場合」という意味になります。
```go
		switch x := err.(type) {
		case interface{ Unwrap() error }:
```

ここで`fmt.Errorf`の実装を見てみます。

```go
func Errorf(format string, a ...any) error {
	p := newPrinter()
	p.wrapErrs = true
	p.doPrintf(format, a)
	s := string(p.buf)
	var err error
	switch len(p.wrappedErrs) {
	case 0:
		err = errors.New(s)
	case 1:
		w := &wrapError{msg: s}
		w.err, _ = a[p.wrappedErrs[0]].(error)
		err = w
	default:
		if p.reordered {
			slices.Sort(p.wrappedErrs)
		}
		var errs []error
		for i, argNum := range p.wrappedErrs {
			if i > 0 && p.wrappedErrs[i-1] == argNum {
				continue
			}
			if e, ok := a[argNum].(error); ok {
				errs = append(errs, e)
			}
		}
		err = &wrapErrors{s, errs}
	}
	p.free()
	return err
}
```

色々書いてはありますが、要は`fmt.Errorf`は第二引数以降が存在していれば`&wrapErrors`を返すという意味になります。
この`wrapErrors`は以下のように定義されています。

```go
type wrapErrors struct {
	msg  string
	errs []error
}

func (e *wrapErrors) Error() string {
	return e.msg
}

func (e *wrapErrors) Unwrap() []error {
	return e.errs
}
```

`fmt.Errorf`で生成されたエラーの実体は`wrapErrors`であり、`Unwrap`メソッドを持っているということになります。
以下の分岐内によって、errorは`Unwrap`され、`nil`でなければ分岐を抜けます。

```go
		switch x := err.(type) {
        case interface{ Unwrap() error }:
            err = x.Unwrap()
            if err == nil {
                return false
            }
}
```

この処理は`for`内にあるため、**第一引数の`err`は`Unwrap`を繰り返され、第二引数の`target`とポインタが一致しているかを、繰り返し評価されます。**
この繰り返しは第一引数の`err`が`Unwrap`できなくなるか、`Unwrap`したら`nil`になるまで、続きます。

Q3において、`fmt.Errorf`で生成されたエラーは`Unwrap`されて、`ErrSentinel`が出てきます。
err2は同じく`ErrSentinel`を指しているので、`true`が返ります。

```go
    err1 = fmt.Errorf("wrapped: %w", ErrSentinel)
    err2 = ErrSentinel
    fmt.Printf("Q3: %v\n", errors.Is(err1, err2)) // true
```

## Q4

```go
    err1 = fmt.Errorf("wrapped: %w", ErrSentinel)
    err2 = fmt.Errorf("wrapped: %w", ErrSentinel)
    fmt.Printf("Q4: %v\n", errors.Is(err1, err2))
```

Q3と同様に、`fmt.Errorf`で生成されたエラーは`Unwrap`されて、`ErrSentinel`が出てきます。
しかし、err2は`fmt.Errorf`で生成されたエラーであり、`err1`や`ErrSentinel`とは異なるポインタを持っています。
よって、`false`が返ります。

```go
    err1 = fmt.Errorf("wrapped: %w", ErrSentinel)
    err2 = fmt.Errorf("wrapped: %w", ErrSentinel)
    fmt.Printf("Q4: %v\n", errors.Is(err1, err2)) // false
```

## Q5

```go
    err1 = ErrSentinel
    err2 = fmt.Errorf("wrapped: %w", ErrSentinel)
    fmt.Printf("Q5: %v\n", errors.Is(err1, err2))
```

もう一度`errors.is`の実装を覗いてみます。

```go
func is(err, target error, targetComparable bool) bool {
	for {
        // 中略
		switch x := err.(type) {
		case interface{ Unwrap() error }:
			err = x.Unwrap()
			if err == nil {
				return false
			}
        // 中略
		default:
			return false
		}
	}
}
```

注意しなくてはいけないのが、`Unwrap`されるのは第一引数の`err`の方であり、**第二引数の`target`は`Unwrap`されない**ということです。

Q5のコードでは`ErrSentinel`が第一引数にきており、第二引数である`fmt.Errorf`で生成されたエラーは`Unwrap`されません。
よって、`false`が返ります。

```go
    err1 = ErrSentinel
    err2 = fmt.Errorf("wrapped: %w", ErrSentinel)
    fmt.Printf("Q5: %v\n", errors.Is(err1, err2)) // false
```

## Q6
    
```go
    err1 = &MyError{Msg: ErrMsg}
    err2 = errors.New(ErrMsg)
    fmt.Printf("Q6: %v\n", errors.Is(err1, err2))
```

もう一度`errors.is`の実装を覗いてみます。

```go
func is(err, target error, targetComparable bool) bool {
for {
    // 中略
    if x, ok := err.(interface{ Is(error) bool }); ok && x.Is(target) {
        return true
    }
	// 中略
}
```

これは要は`bool`を返す`Is`メソッドを持っているか、持っていればその`Is`メソッドで評価するという意味です。

`MyError`は以下のように定義されています。

```go
type MyError struct {
    Msg string
}
// 中略
func (e *MyError) Is(target error) bool {
    return e.Msg == target.Error()
}
```

`MyError`は`Is`メソッドを持っており、その中でエラーメッセージの一致を評価しています。
`err1`、`err2`はそれぞれ異なるエラーの生成方法で生成されていますが、`MyError`の`Is`で評価され、`Is`内で評価対象となるエラーメッセージが一致しているため、`true`が返ります。

```go
    err1 = &MyError{Msg: ErrMsg}
    err2 = errors.New(ErrMsg)
    fmt.Printf("Q6: %v\n", errors.Is(err1, err2)) // true
```

## Q7

```go
    err1 = ErrSentinel
    err2 = &MyError{Msg: ErrMsg}
    fmt.Printf("Q7: %v\n", errors.Is(err1, err2))
```

Q5の解説において、以下のような注意を述べました。
> 注意しなくてはいけないのが、`Unwrap`されるのは第一引数の`err`の方であり、**第二引数の`target`は`Unwrap`されない**ということです。

`errors.is`をもう一度確認します。
この注意は`err.(interface{ Is(error) bool })`の部分に関しても同様です。
すなわち、**第一引数の`err`の`Is`メソッドだけを見ており、第二引数の`target`の`Is`メソッドは評価されません。**

```go
func is(err, target error, targetComparable bool) bool {
for {
    // 中略
    if x, ok := err.(interface{ Is(error) bool }); ok && x.Is(target) {
        return true
    }
	// 中略
}
```

Q7のコードでは`err1`が`ErrSentinel`であり、これは`errors.New`で生成されたエラーです。
`errors.New`で生成されたエラーには`Is`メソッドがなく、`MyError`の`Is`メソッドは評価されないため、`false`が返ります。

```go
    err1 = ErrSentinel
    err2 = &MyError{Msg: ErrMsg}
    fmt.Printf("Q7: %v\n", errors.Is(err1, err2)) // false
```

# 解答

```
Q1: true
Q2: false
Q3: true
Q4: false
Q5: false
Q6: true
Q7: false
```

# まとめ再掲

- `errors.Is`の挙動を理解したいなら、`errors`パッケージの`is`関数を見よ
- `is`関数では基本的にはポインタが一致してるかだけを見る
- 基本的にはエラーメッセージの中身を比較するような実装にはなっていない
- `errors.Is`の第一引数の`err`は`Unwrap`を繰り返され、第二引数とポインタが一致しているかが見られる
- `errors.Is`で`Unwrap`を繰り返されるのは第一引数の方であり、第二引数の`target`は`Unwrap`されない
- `errors.Is`で第一引数の`err`に`Is`メソッドがある場合は、その返り値で評価される
- `errors.Is`で`Is`メソッドが評価されるのは第一引数の方であり、第二引数の`target`の`Is`メソッドは評価されない